### PR TITLE
Test boost 1.68.0

### PR DESCRIPTION
**Pull request by BenjaminW3**

### CODE REVIEW

**Comment by BenjaminW3** [comment]
There seems to be a problem with the nvcc version number detection. Investigation is running.

**Comment by BenjaminW3** [comment]
The nvcc compiler detection has been added in alphabetical order into the other compilers. Therefore gcc and clang are detected earlier. nvcc emulates the underlying host compiler and is therefore first detected as the host compiler itself. After this, `BOOST_COMP_NVCC_EMULATED` is set to the correct version instead of `BOOST_COMP_NVCC`, which is only defined to zero.
This can be fixed by explicitly including `<boost/predef/compiler/nvcc.h>` prior to any other predef header.
I am unsure how this should be fixed in boost predef. Chaning the order of detection includes should help, but then we would destroy the alphabetical order.

**Comment by ax3l** [comment]
> I am unsure how this should be fixed in boost predef. 

Sounds like a bug to me.

**Comment by BenjaminW3** [comment]
The implementations of those compiler detection headers are all nearly identical and the structure is predefined. I found that boost config also moved nvcc to the front so I will create a boost predef pull request.

File: include/alpaka/core/BoostPredef.hpp [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        #else
+            #define BOOST_COMP_NVCC BOOST_VERSION_NUMBER(__CUDACC_VER_MAJOR__, __CUDACC_VER_MINOR__, __CUDACC_VER_BUILD__)
+        #endif
+    #else
+        #define BOOST_COMP_NVCC BOOST_VERSION_NUMBER_NOT_AVAILABLE
+    #endif
+#endif
+
+//-----------------------------------------------------------------------------
+#if BOOST_COMP_INTEL == 0 // Work around for broken intel detection
```
[the remark this answers is not in the document]
Comment (BenjaminW3) [reply include/alpaka/core/BoostPredef.hpp]: I also had a look at this and there are still two open tickets for intel in predef so I think it it is still broken. We might want to find the original author of this code and the corresponding ticket.

[the remark this answers is not in the document]
Comment (ax3l) [reply include/alpaka/core/BoostPredef.hpp]: I think @theZiz added this :)

[the remark this answers is not in the document]
Comment (BenjaminW3) [reply include/alpaka/core/BoostPredef.hpp]: it has been added in #299. Still have to find out the boost version.

[the remark this answers is not in the document]
Comment (BenjaminW3) [reply include/alpaka/core/BoostPredef.hpp]: It should have landed in boost 1.64.0

[the remark this answers is not in the document]
Comment (BenjaminW3) [reply include/alpaka/core/BoostPredef.hpp]: I have added a comment about the boost version to this workaround

**Comment by BenjaminW3** [comment]
Ready for review!

File: include/alpaka/core/BoostPredef.hpp [hunk 2]
```diff
@@ -0,0 +1,99 @@
+/**
+ * \file
+ * Copyright 2017 Alexander Matthes
```
Comment (ax3l) [inline include/alpaka/core/BoostPredef.hpp]: You might want to update the authors to `Alexander Matthes, Benjamin Worpitz`.

File: include/alpaka/core/BoostPredef.hpp [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        #define BOOST_ARCH_PTX BOOST_VERSION_NUMBER_NOT_AVAILABLE
+    #endif
+#endif
+
+//-----------------------------------------------------------------------------
+// In boost since 1.68.0
+// nvcc CUDA compiler detection
+
+#include <boost/version.hpp>
+#if BOOST_VERSION >= 106800
```
Comment (ax3l) [inline include/alpaka/core/BoostPredef.hpp]: use `==`. It should be fixed in the next release if we provide an upstream PR now.
maybe also link the issue/PR here

Comment (BenjaminW3) [reply include/alpaka/core/BoostPredef.hpp]: As long as we do not have an accepted solution in boost it is more future proof with `>=`. I am still thinking about the correct solution in boost/predef and this PR can be merged before this is solved.

File: include/alpaka/core/BoostPredef.hpp (hunk 2 above)
Comment (BenjaminW3) [reply include/alpaka/core/BoostPredef.hpp]: done

**Comment by BenjaminW3** [comment]
Push

**Comment by ax3l** [comment]
@BenjaminW3 did you by coincidence open an upstream PR to boost already?

**Comment by BenjaminW3** [comment]
https://github.com/boostorg/predef/pull/84

**Pull request merged**